### PR TITLE
Update VirtualModulesPlugin from upstream

### PR DESCRIPTION
**Pull request by Smittyvb**
tl;dr: Webpack 5 support, fixes `TypeError: Cannot read property 'data' of undefined`

This updates the internal VirtualModulesPlugin, which is copied from [`webpack-virtual-modules`](https://github.com/sysgears/webpack-virtual-modules), which was originally added by @kaisermann. These changes allow for using Webpack 5 with `svelte-loader`. The changes of #125, which have also been fixed upstream, have been preserved. See https://github.com/sveltejs/svelte-loader/pull/59#issuecomment-385381444 for the background on why the source code is just copied over instead of added as a dependency. Due to the `svelte-loader` specific changes already in `virtual.js`, I have manually gone through the commit history of [`webpack-virtual-modules`](https://github.com/sysgears/webpack-virtual-modules) and applied relevant changes instead of just copy-pasting the new source code. There is also a small change to `index.js` to prevent the error handling code from itself throwing errors, since that was causing some (now fixed) troubles when debugging my changes.

This also adds in the literal text of the license of [`webpack-virtual-modules`](https://github.com/sysgears/webpack-virtual-modules), to a newly created `LICENSE` file, since "The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software".

Tests and lints pass, and I have verified this works with both Webpack 4 and Webpack 5

[continuation from #132, since GH won't let me re-open it]

Closes https://github.com/sveltejs/svelte-loader/pull/133

### CODE REVIEW

File: lib/virtual.js [hunk 1]
```diff
@@ -1,9 +1,26 @@
+// Adapted from https://github.com/sysgears/webpack-virtual-modules
+
 var VirtualStats = require('./virtual-stats');
+var path = require("path");
```
Comment (benmccann) [inline lib/virtual.js:4]: we should use single quotes here to match the line above

File: index.js [hunk 2]
```diff
@@ -178,6 +178,6 @@ module.exports = function(source, map) {
 	}, err => callback(err)).catch(err => {
 		// wrap error to provide correct
 		// context when logging to console
-		callback(new Error(`${err.name}: ${err.toString()}`));
+		callback(new Error(`${err.name || "Error"}: ${err.toString()}`));
```
Comment (benmccann) [inline index.js:181]: we should probably put Error in single quotes

File: lib/virtual.js [hunk 3]
```diff
@@ -75,14 +146,81 @@ VirtualModulesPlugin.prototype.writeModule = function(filePath, contents) {
...
+}
+
+function getReadDirBackend(fileSystem) {
+	if (fileSystem._readdirBackend) {
+		return fileSystem._readdirBackend;
+	} else if (fileSystem._readdirStorage) {
+		return fileSystem._readdirStorage;
 	} else {
-		storage.data[key] = value;
+		throw new Error("Couldn't find a readDirStorage from Webpack Internals");
```
Comment (benmccann) [inline lib/virtual.js:223]: should probably put this string in single quotes

**Comment by Smittyvb** [comment]
@benmccann I've changed all uses of double quotes to single quotes.

File: lib/virtual.js [hunk 4]
```diff
@@ -75,14 +146,81 @@ VirtualModulesPlugin.prototype.writeModule = function(filePath, contents) {
...
+}
+
+function getReadDirBackend(fileSystem) {
+	if (fileSystem._readdirBackend) {
+		return fileSystem._readdirBackend;
+	} else if (fileSystem._readdirStorage) {
+		return fileSystem._readdirStorage;
 	} else {
-		storage.data[key] = value;
+		throw new Error('Couldn\'t find a readDirStorage from Webpack Internals');
```
Comment (benmccann) [inline lib/virtual.js:223]: I probably shouldn't have commented on this line. I missed that it has an apostrophe that needed to be escaped. We use double quotes when there's an apostrophe to avoid the need to escape. We've setup an [eslint config](https://github.com/sveltejs/eslint-config) on the main svelte repo to check for this stuff, but haven't set it up on this project yet, so sorry it's not quite as automated here yet

**Review by benmccann — APPROVED** [review]
lgtm

**Comment by larixer** [comment]
@Smittyvb You might probably need to check whether you affected by this Issue:
https://github.com/sysgears/webpack-virtual-modules/issues/68
and related PR:
https://github.com/sysgears/webpack-virtual-modules/pull/69

**Comment by Smittyvb** [comment]
I've implemented https://github.com/sysgears/webpack-virtual-modules/pull/69/commits/489401990228b94faee4df44fb85066acc89d24a from upstream, which implements watch mode support in a way that is compatible with Webpack 5. So to recap:
- Currently: no watch mode support
- #133 (superseded by this PR): Webpack 4 watch mode support
- This PR: Webpack 4+5 watch mode support

File: lib/virtual.js [hunk 5]
```diff
@@ -16,24 +37,69 @@ function VirtualModulesPlugin(compiler) {
...
+			const statStorage = getStatStorage(this);
+			const fileStorage = getFileStorage(this);
+			const readDirStorage = getReadDirBackend(this);
 			this._virtualFiles = this._virtualFiles || {};
 			this._virtualFiles[file] = { stats: stats, contents: contents };
-			setData(this._statStorage, file, [null, stats]);
-			setData(this._readFileStorage, file, [null, contents]);
+			setData(statStorage, file, createWebpackData(stats));
+			setData(fileStorage, file, createWebpackData(contents));
+			var segments = file.split(/[\\/]/);
```
Comment (benmccann) [inline lib/virtual.js:58]: I generally prefer `const` and `let` over `var` (same below)

File: lib/virtual.js [hunk 6]
```diff
@@ -12,28 +33,73 @@ function VirtualModulesPlugin(compiler) {
...
+					const timeMs = time.getTime();
+					const dirStats = new VirtualStats({
+						dev: 8675309,
+						nlink: 0,
+						uid: 1000,
+						gid: 1000,
+						rdev: 0,
+						blksize: 4096,
+						ino: inode++,
+						mode: 16877,
```
Comment (benmccann) [inline lib/virtual.js:76]: why is mode different here than line 126? can we refactor to share the same `VirtualStats` or at least `VirtualStats` options?

File: lib/virtual.js [hunk 7]
```diff
@@ -12,28 +33,73 @@ function VirtualModulesPlugin(compiler) {
...
 			this._virtualFiles = this._virtualFiles || {};
 			this._virtualFiles[file] = { stats: stats, contents: contents };
-			setData(this._statStorage, file, [null, stats]);
-			setData(this._readFileStorage, file, [null, contents]);
+			setData(statStorage, file, createWebpackData(stats));
+			setData(fileStorage, file, createWebpackData(contents));
+			const segments = file.split(/[\\/]/);
+			let count = segments.length - 1;
+			const minCount = segments[0] ? 1 : 0;
+			while (count > minCount) {
```
Comment (benmccann) [inline lib/virtual.js:61]: what is this `while` looping doing? and why is it newly required but seems to also apply to webpack 4? a code comment here would be helpful

File: lib/virtual.js [hunk 8]
```diff
@@ -74,15 +140,104 @@ VirtualModulesPlugin.prototype.writeModule = function(filePath, contents) {
...
+		});
+	}
 };
 
-function setData(storage, key, value) {
-	if (storage.data instanceof Map) {
-		storage.data.set(key, value);
+function getData(storage, key) {
+	// Webpack 5
+	if (storage._data instanceof Map) {
```
Comment (benmccann) [inline lib/virtual.js:179]: you could probably simplify both this method and `setData` a lot by creating an intermediate `data` variable:
```
const data = storage._data /* webpack 5 */ || storage.data /* webpack 4 */;
if (data instanceof Map) {
  return storage.data.get(key);
}
return data.data[key];
```

**Comment by larixer** [comment]
@Smittyvb Guys, if you are up to it, I would be open to PR to expose some interface inside `webpack-virtual-modules` to make it easier for you to consume it directly inside `svelte-loader`, it might make future upgrades easier. Please consider this as an option now or in the future.

File: lib/virtual.js (hunk 7 above)
Comment (Smittyvb) [reply lib/virtual.js:88]: This is from https://github.com/sysgears/webpack-virtual-modules/commit/a73d53f8775ab3e5dd2bc57cc5e18d63408aefc8, it creates directories that the file is in.

**Comment by Smittyvb** [comment]
@larixer I'm working on something like that currently: https://github.com/Smittyvb/svelte-loader/tree/wvm-as-dep

**Comment by benmccann** [comment]
That would be great and I think would help get this PR checked in. I talked to some of the other maintainers about getting this PR checked in since I don't have much webpack background and haven't worked on this repo much, and I think there was some feeling that they'd like to see webpack 5 support be a bit cleaner/simpler. I'm not sure exactly how you would do that, but if there's a nice interface for `webpack-virtual-modules` to support this that sounds like it could help

**Comment by jquense** [comment]
@larixer would also love to see something like that. I've forked the plugin a few times in order to allow bootstrapping in a loader. 

if you are looking for ideas on an interface i've found this effective: https://github.com/4Catalyzer/astroturf/blob/v1/src/VirtualModulePlugin.ts#L99-L108

File: lib/virtual.js [hunk 9]
```diff
@@ -50,39 +122,100 @@ function VirtualModulesPlugin(compiler) {
...
+	}
 };
 
-function setData(storage, key, value) {
-	if (storage.data instanceof Map) {
-		storage.data.set(key, value);
+function getData(storage, key) {
+	const data = storage._data /* webpack 5 */ || storage.data /* webpack 4 */;
+	if (data instanceof Map) {
+		return storage.data.get(key);
```
Comment (lindenquan) [inline lib/virtual.js:164]: This should be `data.get(key)`

Comment (martinoppitz) [reply lib/virtual.js:164]: Habe zu fix data.get

File: lib/virtual.js [hunk 10]
```diff
@@ -50,39 +122,100 @@ function VirtualModulesPlugin(compiler) {
...
+	}
+	return data.data[key];
+}
+
+function setData(backendOrStorage, key, valueFactory) {
```
Comment (martinoppitz) [inline lib/virtual.js:169]: This implementation is not nice. Better looks like getData, a webpack switch and a generic code

**Review by martinoppitz — COMMENTED** [review]
Code advices

**Comment by martinoppitz** [comment]
Oh sorry - I will fix the bug - if I can,

**Comment by martinoppitz** [comment]
Hello @Smittyvb, I fixed it. Please merge it again https://github.com/Smittyvb/svelte-loader/pull/3

**Comment by olexlyap** [comment]
Hello everyone, it looks like there is something wrong with style handling on this branch when `svelte-preprocess` is enabled. I get errors like this for all of my components which have styles:
```
Module not found: Error: Can't resolve '/path-to-my-app/src/components/Header/index.svelte.css' in '/path-to-my-app/src/components/Header'
```

The directory structure for the `Header` component (and all others) is as follows:
```
Header
  - index.scss
  - index.svelte
```
And the style is connected to the component like this: `<style type="text/scss" src="index.scss"></style>`.

Here is my Webpack configuration:
```js
{
  entry, // custom generated list of entry points
  mode: 'production',
  output: {
    path: path.join(projectRoot, 'dist'),
    filename: '[name].js',
  },
  resolve: {
    alias: {
      svelte: path.resolve(projectRoot, 'node_modules', 'svelte'),
    },
    extensions: ['.mjs', '.js', '.svelte', '.css'],
    mainFields: ['svelte', 'browser', 'module', 'main'],
  },
  module: {
    rules: [
      { test: /.m?js$/, type: 'javascript/auto' },
      {
        test: /\.svelte$/,
        use: {
          loader: 'svelte-loader',
          options: {
            hotReload: false,
            hydratable: true,
            preprocess: sveltePreprocess(), // imported from 'svelte-preprocess'
            dev: false,
            emitCss: true,
          },
        },
      },
      {
        test: /\.css$/,
        use: [MiniCssExtractPlugin.loader, 'css-loader'],
      },
    ],
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: '[name].css',
    }),
  ],
  optimization: {
    splitChunks: {
      chunks: 'all',
    },
  },
}
```
What fixes the error:
- using the latest version from npm or the `master` branch from this repo and downgrading to Webpack 4
- removing the `preprocess` option and converting the styles into CSS

What doesn't fix the error:
- converting the styles into CSS
- switching between Webpack v4 and v5

**Comment by martinoppitz** [comment]
I think the changed files have nothing to do with CSS. The misconduct with regard to webpack 5 and preprocess is a separate topic.

It would be cool if the PR comes so that Webpack5 can work in less complex situations.

**Comment by martinoppitz** [comment]
the same story here:
- https://github.com/vuetifyjs/vuetify-loader/issues/136 and here
- https://github.com/aurelia/aurelia/issues/982 and here
- https://github.com/sveltejs/svelte-loader/pull/136 and so on

Many thanks to Angular for influencing Webpack and the unconsolidated upgrade to v5.

**Comment by deleonio** [comment]
That is solved for us leanupjs/leanup#37.

**Comment by Smittyvb** [comment]
Superseded by #146, which fixes the problem in a _much_ better and more maintainable manner.
